### DISCUSSION TITLE
how to romote debug skywalking-java#apm-agent module

### DISCUSSION THREAD

**Discussion by yuanjinzhong** [apache/skywalking#11835] [Q&A] [answered]
i had follw this article :https://skywalking.apache.org/zh/2019-01-24-skywalking-remote-debug/
and learn how to romote debug **skywalking-java#<font color=red>apm-agent</font> module**

blow is my step:
* step one :edit my micro service configuration(Specify parameter of -javaagent and remote debug ) and start debug,
![image](https://github.com/apache/skywalking/assets/13908993/a020e7d9-23a3-4fd2-ace0-134cb42c1f36)
* step two : set remote jvm debug ,and select  **apm-agent** module, and start debug
![image](https://github.com/apache/skywalking/assets/13908993/617dfb6a-8200-4a38-82d2-c68e027cb22b)

but there appere  a error,   
![image](https://github.com/apache/skywalking/assets/13908993/a013f09a-6d04-430f-9ec8-201563ee5211)

<font color=red>"Unable to open debugger port (localhost:5005): java.net.ConnectException "Connection refused""</font>

any one has some idea to help me ?  thanks

**Comment by wu-sheng** [comment]
QQ groups are gone, we are not going to maintain that.

If you need to discuss in Chinese, you could learn how to join slack CN channel. https://juejin.cn/post/7262010787228074042

**Reply by yuanjinzhong** [reply]
i get it; by the way,  boss wu, does you have some idea of my question ? thanks

**Reply by wu-sheng** [reply]
I didn't debug agents for years. But it is a general way for Java, nothing special.

**Reply by yuanjinzhong** [reply]
i get it ,thanks boss wu

**Reply by yuanjinzhong** [reply]
boss,may i unmark this reply as the answer, then some other boy knowed this question  can help me ?

**Reply by wu-sheng** [reply]
Go ahead, that's your choice.

**Reply by yuanjinzhong** [reply]
thanks

**Comment by revolyw** [comment]
@yuanjinzhong 
You can try moving -agentlib:jdwp... behind -javaagent:...skywalking.agent.jar and change the parameter value of suspend from n to y. 

This will cause the program being debugged to pause until the remote debug connection is successful.

**Reply by yuanjinzhong** [reply]
thanks bro!  i will try later; 

by the way you also tell me a new knowlege of parameter **suspend**, which i never know ; thanks !

**Reply by yuanjinzhong** [reply]
@revolyw  
bro,  I tried that. It didn't work

when i start my micro service,the port 5005,was not listened ?

![image](https://github.com/apache/skywalking/assets/13908993/9fd9208f-e398-4c80-9fb8-4b8d88d95e76)

* **use telnet to test:**

![image](https://github.com/apache/skywalking/assets/13908993/4010b032-8595-457d-bdc6-58db38e0e698)

* **use lsof to test**

![image](https://github.com/apache/skywalking/assets/13908993/ccc87ac8-8b2e-4bb6-bd12-73c5b5e234b9)

**Reply by revolyw** [reply]
It still seems to be a matter of parameters. Change the parameter value of address from *:5005 to 5005 and try again

**Reply by revolyw** [reply]
This is my vm options.  fyi.
![image](https://github.com/apache/skywalking/assets/10175669/a4fb89f3-dfd6-48cb-9f6e-c447ded608f9)

**Reply by yuanjinzhong** [reply, marked as answer]
thanks bro, i try a new way instead of jvm remote debug, reference this article : **https://juejin.cn/post/7238236506854604837**

in my micro service Idea windows, import skywalking-agent as a new module, an start the micro service with the specify parameter:

**-javaagent:/Users/yuanjinzhong/Documents/java/tools/skywalking/apache-skywalking-java-agent-8.9.0/skywalking-agent/skywalking-agent.jar** , and it work

![image](https://github.com/apache/skywalking/assets/13908993/89c51bcb-b874-4790-b0ef-7afba3d645b0)
![image](https://github.com/apache/skywalking/assets/13908993/cd28074d-53a9-4cc7-95b7-63338bd3429f)
![image](https://github.com/apache/skywalking/assets/13908993/b97685d9-d05f-4530-b9f2-4b9e8390fb66)

**Reply by revolyw** [reply]
Cool. It solves your problem and is the most efficient debugging method for a single project.
